From 269b96d76e91b6e0d793d812a60f988c3b49ca25 Mon Sep 17 00:00:00 2001
From: Usuario de Programacion <pairprog@pairprog06.(none)>
Date: Wed, 18 Feb 2009 17:32:01 +0100
Subject: [PATCH] Changed the returned simplified type when DB type is XML.
 This way the schema is correctly dumped to schema.rb

---
 .../connection_adapters/postgresql_adapter.rb      |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
index 913bb52..ce7d291 100644
--- a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
@@ -99,7 +99,7 @@ module ActiveRecord
               :string
             # XML type
             when /^xml$/
-              :string
+              :xml
             # Arrays
             when /^\D+\[\]$/
               :string              
-- 
1.5.4.3


From 9aaa72e09465be2ccd996a6da8f957ec21027060 Mon Sep 17 00:00:00 2001
From: Usuario de Programacion <pairprog@pairprog06.(none)>
Date: Wed, 18 Feb 2009 17:45:39 +0100
Subject: [PATCH] Added short hand column definition for migrations

---
 .../connection_adapters/postgresql_adapter.rb      |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
index ce7d291..2285c02 100644
--- a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
@@ -39,6 +39,12 @@ module ActiveRecord
   end
 
   module ConnectionAdapters
+    class TableDefinition
+      def xml(*args)
+        options = args.extract_options!
+        column(args[0], :xml, options)
+      end
+    end
     # PostgreSQL-specific extensions to column definitions in a table.
     class PostgreSQLColumn < Column #:nodoc:
       # Instantiates a new PostgreSQL column definition in a table.
-- 
1.5.4.3


From fde1e1837721e43b5721936ed9df36f53d81b11e Mon Sep 17 00:00:00 2001
From: Usuario de Programacion <pairprog@pairprog06.(none)>
Date: Thu, 19 Feb 2009 10:29:16 +0100
Subject: [PATCH] Added fallback method to handle cases where the database doesn't have a XML datatype

---
 .../abstract/schema_definitions.rb                 |   15 +++++++++++++++
 1 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
index 24c734c..121b8c0 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
@@ -316,6 +316,21 @@ module ActiveRecord
         @base = base
       end
 
+      #Handles non supported datatypes - e.g. XML
+      def method_missing(symbol, *args)
+        if symbol.to_s == 'xml'
+	  xml_column_fallback(args)
+        end
+      end
+
+      def xml_column_fallback(*args)
+	case @base.adapter_name.downcase
+	  when 'sqlite', 'mysql'
+            options = args.extract_options!
+            column(args[0], :text, options)
+	end       
+      end
+
       # Appends a primary key definition to the table definition.
       # Can be called multiple times, but this is probably not a good idea.
       def primary_key(name)
-- 
1.5.4.3

